shared/src/main/scala/async/futures.scala

```diff
@@ -261,23 +221,67 @@ object Future:
 
   end extension
 
-  /** A promise defines a future that is be completed via the promise's `complete` method.
+  /** A promise defines a future that is be completed via the `complete` method.
     */
-  class Promise[T]:
-    private val myFuture = new CoreFuture[T]:
-      fut =>
-      override def cancel(): Unit =
-        if setCancelled() then fut.complete(Failure(new CancellationException()))
+  trait Promise[T] extends Future[T]:
+    inline def asFuture: Future[T] = this
+
+    /** Define the result value of `future`. */
+    def complete(result: Try[T]): Unit
 
-    /** The future defined by this promise */
-    val future: Future[T] = myFuture
+  object Promise:
+    def apply[T](): Promise[T] =
+      new CoreFuture[T] with Promise[T]:
+        override def cancel(): Unit =
+          if setCancelled() then complete(Failure(new CancellationException()))
 
-    /** Define the result value of `future`. However, if `future` was cancelled in the meantime complete with a
-      * `CancellationException` failure instead.
-      */
-    def complete(result: Try[T]): Unit = myFuture.complete(result)
+        /** Define the result value of `future`. However, if `future` was cancelled in the meantime complete with a
+          * `CancellationException` failure instead.
+          */
+        override def complete(result: Try[T]): Unit = super[CoreFuture].complete(result)
```

Comment:
Interesting. So this is fine because we are overriding the public `complete` from `Promise` (and not from `CoreFuture`), and so it's completely fine to keep it public...?

Reply:
No, it's just valid code to increase visibility as in
```scala
object X:
  class A:
    private[X] def foo() = 1

  class B extends A:
    override def foo() = super.foo()

@main def bar() = println(X.B().foo())
```

Also see https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Object.html#clone() for that design in Java Standard Library.